### DISCUSSION TITLE
Automatic aggregated metrics

### DISCUSSION THREAD

**Discussion by JuN12456** [ClickHouse/ClickHouse#60135] [Q&A]
Hello, i'm new at using clickhouse, so sorry if my question is dumb but i haven't find what i'm searching in clickhouse documentation. 

I possess data spanning over two years, comprising billions of rows, which understandably results in my query being very slow and consume a considerable amount of RAM.
I'm interested in exploring if there's a method, perhaps through the use of a temporary table or another ClickHouse functionality, to automatically aggregate this data daily. This would ideally allow me to work with a much smaller table that contains only the daily results of my query.
I haven't test yet if adding _name_ in my order by when creating my table will speed up my query, but in anycase it will not help with the ram consumption. 

i have  the following table :
```

CREATE TABLE events (
    ad_unit_id String ,
    date DateTime ,
    name String ,
    user Nullable(String),
    country Nullable(String),
    device Nullable(string) 

ENGINE = MergeTree()
ORDER BY (date)
PARTITION BY ((toYYYYMMDD(date), name));

```

i'm executing a query like this one:

```
select
    toDate(date) as day,
    country ,
    device ,
    count(distinct user) 
from
    events
where
    name = 'search'
    and device = 'xxx'
group by
        day,
        country 
```

Thanks for your help :) .

**Comment by ArctypeZach** [comment]
Hi @JuN12456 -

You can try using projections: https://clickhouse.com/docs/en/sql-reference/statements/alter/projection
Some additional reading: https://clickhouse.com/blog/clickhouse-faster-queries-with-projections-and-primary-indexes

Another small nit: you can use `uniq(user)` or `uniqExact(user)` instead of `count(distinct user)`.  `uniq()` will perform better but is not guaranteed to be exact.
